src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +92,221 @@ object RuleUtils {
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param originalPlan original plan
+   * @param indexPlan replaced plan with index
+   * @return complementIndexPlan integrated plan of indexPlan and complementPlan
+   */
+  private def getComplementIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan,
+      indexPlan: LogicalPlan): LogicalPlan = {
+    // 1) Replace the location of LogicalRelation with appended files
+    val complementIndexPlan = originalPlan transformDown {
+      case baseRelation @ LogicalRelation(
+            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion
+        val updatedOutput =
+          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
+
+        val filesAppended = (location.allFiles
+          .map(FileInfo(_))
+          .toSet -- index.allSourceFiles).toSeq.map(f => new Path(f.name))
+
+        if (filesAppended.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
+    }
+
+    if (!originalPlan.equals(complementIndexPlan)) {
+      // Remove sort order because we cannot guarantee the ordering of source files
+      val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+
+      def getIndexedAttrs(
+          plan: LogicalPlan,
+          indexedColumns: Seq[String]): Seq[Option[Attribute]] = {
+        val attrMap = plan.output.attrs.map(attr => (attr.name, attr)).toMap
+        indexedColumns.map(colName => attrMap.get(colName))
+      }
+
+      // 2) Perform on-the-fly Shuffle with the same partition structure of index
+      // so that we could avoid incurring Shuffle of whole index data at merge stage.
+      // In order to utilize push-down filters, we would locate Shuffle node after
+      // Project or Filter node. (Case 1)
+      // However, if Project node excludes any of indexedColumns, Shuffle will be
+      // converted to RoundRobinPartitioning which can cause wrong result issues.
+      // So Shuffle should be located before Project node in that case. (Case 2)
+      // Case 1) Shuffle => Project => Filter => Relation (Project&Filter will be pushed down)
+      // Case 2) Project => Shuffle => Filter => Relation (Filter will be pushed down)
+      var shuffleInjected = false
+      val shuffled = complementIndexPlan transformDown {
```

Comment:
@ reviewers, please let me know if you have any good idea or suggestion for this transformDown block 😢

Reply:
Do you mean how to refactor this?

You could try an extractor. Maybe something like the following?
```scala
val shuffled = complementIndexPlan transformDown {
  case p if shuffleInjected => p
  case ExtractTopLevelPlanForShuffle(plan, indexedAttrs) =>
     shuffleInjected = true
     RepartitionByExpression(indexedAttrs.flatten, plan, index.numBuckets)
}
```

Reply:
If you want to avoid calling `getIndexedAttrs` twice:
```scala
val shuffled = complementIndexPlan transformDown {
  case p if shuffleInjected => p
  case ExtractTopLevelPlanForShuffle(plan, indexedAttrs, requireBlah)
    if !requireBlah || indexedAttrs.forall(_.isDefined) =>
      shuffleInjected = true
      RepartitionByExpression(indexedAttrs.flatten, plan, index.numBuckets)
}
```

Reply:
Thank you:) It's very helpful 👍 👍